File: Ejercicio1_sin_funciones.rb
require 'daru'

# escribir el algoritmo
# realizando sus propias funciones (busqueda, minimo, maximo, promedio y correlacion).

# read the file
data = Daru::DataFrame.from_csv('total_cases.csv')
# get number of rows, shape return array with number of rows and ncols
nrows = data.shape[0] - 1
# get vector from dataframe last row (last date)
df_total_cases = data.row[nrows]

# remove unused columns
arr = ['date', 'World', 'European Union', 'Europe', 'Asia', 'North America', 'South America', 'Africa', 'Oceania']
arr.each do |i|
  df_total_cases.delete_at(i)
end
df_total_cases.rename 'Casos'

# ---------------------
# 4 smallest countries + py
# ---------------------

# sort asc
df_smallest = df_total_cases.sort
df_smallest.rename 'Casos'
# get first 4 elements
df_smallest = df_smallest.head(4)
# concat py value
df_smallest.concat(df_total_cases['Paraguay'], 'Paraguay')
# print
puts 'Los 4 países con menores valores de total de casos positivos. + Py'
puts df_smallest.inspect

# ---------------------
# 4 largest countries + py
# ---------------------

# sort descendent
df_highest = df_total_cases.sort(ascending: false)
df_highest.rename 'Casos'
# get first 4 elements
df_highest = df_highest.head(4)
# concat py value
df_highest.concat(df_total_cases['Paraguay'], 'Paraguay')
# print
puts 'Los 4 países con mayores valores de total de casos positivos. + Py'
puts df_highest.inspect

# ---------------------
# cases per million section
# ---------------------

# read csv
locations = Daru::DataFrame.from_csv('locations.csv')
locations = locations['population']
# new vector with population data an index its same as df_total_cases, now population and df_total_cases has same index
population = Daru::Vector.new(locations, index: df_total_cases.index, name: :population)
# operation between vectors to get total cases per million
df_cases_per_million = df_total_cases * 1_000_000 / population
df_cases_per_million.delete_at('International')
df_cases_per_million.rename('Casos por millon')
# round elements
df_cases_per_million = df_cases_per_million.round

# ---------------------
# 4 largest cases per million + py
# ---------------------

# sort descendent
df_highest = df_cases_per_million.sort(ascending: false)
# get first 4 elements
df_highest = df_highest.head(4)
# concat py value
df_highest.concat(df_cases_per_million['Paraguay'], 'Paraguay')
# print
puts 'los 4 países con mayores casos positivos por millón de habitantes. + Py'
puts df_highest.inspect

# ---------------------
# 4 smallest cases per million + py
# ---------------------

# sort asc
df_smallest = df_cases_per_million.sort
# get first 4 elements
df_smallest = df_smallest.head(4)
# concat py value
df_smallest.concat(df_cases_per_million['Paraguay'], 'Paraguay')
# print
puts 'los 4 países con menores casos positivos por millón de habitantes. + Py'
puts df_smallest.inspect

# ---------------------
# find cases per day section
# ---------------------

# clone df , if not any mod to df_for_cases affect data
df_for_cases = data.clone

# diff deletes the first row, so we too
first_row = 0

# delete unused columns
arr.each do |i|
  df_for_cases.delete_vector(i)
end
# clone the structure to a new dataframe
df_cases_per_day = df_for_cases.clone_structure
# delete first row to fit in diff result

df_cases_per_day.delete_row(first_row)

# loop over every column and do diff
columns = df_for_cases.vectors
columns.each do |i|
  df_cases_per_day[i] = df_for_cases[i].diff
end

# ---------------------
# last 10 days average all countries section
# ---------------------

def rounded_mean(arr)
  cant = arr.count
  total = arr.sum
  if cant != 0
    res = total / cant
    res = res.round
  else
    res = 0
  end
  res
end

# apply mean, round and rename
df_average = df_cases_per_day.tail(10).rename('Promedio')
# apply mean
puts 'Promedio diario de casos positivos de los últimos 10 días de todos los países'
# clone structure
df_avg = df_total_cases.clone
# get indexes
columns = df_avg.index.to_a
# iterate over every column
columns.each do |i|
  df_avg[i] = rounded_mean df_average[i]
end
puts df_avg.inspect

# ---------------------
#  correlation, cases per day
# ---------------------

def correlation(x, y)
  # this calculates correlation between to data arrays
  n = x.count
  # Finding the mean of the series x and y
  mean_x = x.sum / n
  mean_y = y.sum / n
  cov = 0
  std_x = 0
  std_y = 0
  range = [*0..n - 1]
  range.each do |i|
    cov += (x[i] - mean_x) * (y[i] - mean_y)
    std_x += (x[i] - mean_x)**2
    std_y += (y[i] - mean_y)**2
  end
  std_x **= 0.5
  std_y **= 0.5
  numerator = cov
  denominator = std_x * std_y
  if denominator != 0
    return numerator / denominator
  else
    return 0
  end
end

# get columns
df_for_corr = df_cases_per_day.tail(15)
columns = df_for_corr.vectors
# clone previous vector structure
df_corr = df_total_cases.clone_structure
df_corr.rename 'Correlacion'
# iterate
columns.each { |i| df_corr[i] = correlation(df_for_corr['Paraguay'], df_for_corr[i]) }
df_corr['Paraguay'] = -1
puts 'Los países que tienen la mayor correlación en los últimos 15 días, con respecto a Paraguay.'
puts df_corr.sort(ascending: false).head(15).inspect

# ---------------------
# Graph
# ---------------------

require 'matplotlib/pyplot'
plt = Matplotlib::Pyplot
# get the countries to plot, cases per day
df_for_plot = df_cases_per_day['Paraguay', 'Norway', 'Lithuania', 'Israel', 'Ireland']
# replace nil with 0
df_for_plot = df_for_plot.replace_values(nil, 0)
# date string
date = "desde: #{data['date'][0]} hasta: #{data['date'][nrows]}"
# array that stores days
x_axis = [*0..nrows - 1]
# begin to plot
plt.suptitle 'Grafico de Lineas'
plt.title date
plt.xlabel 'dias'
plt.ylabel 'casos por dia'
plt.plot x_axis, df_for_plot['Paraguay'].to_a, label: 'Paraguay'
plt.plot x_axis, df_for_plot['Norway'].to_a, label: 'Norway'
plt.plot x_axis, df_for_plot['Israel'].to_a, label: 'Israel'
plt.plot x_axis, df_for_plot['Lithuania'].to_a, label: 'Lithuania'
plt.plot x_axis, df_for_plot['Ireland'].to_a, label: 'Ireland'
plt.legend
plt.show

File: Ejercicio1_sin_funciones.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# escribir el algoritmo
# realizando sus propias funciones (busqueda, minimo, maximo, promedio y correlacion).
# to sort dict
import operator
# Read the file
df_total_cases = pd.read_csv("total_cases.csv", low_memory=False)
# list of anything that isn't a country
arr = ['World', 'European Union', 'Europe', 'Asia', 'North America', 'South America', 'Africa', 'Oceania']
# delete from dataframe
for i in arr:
    df_total_cases = df_total_cases.drop(columns=i)
data = df_total_cases
data = data.drop(columns='date')
number_rows = data.index.stop - 1
# get subset dataframe for last row (last date)
df = data.iloc[number_rows]
# get number of total cases from paraguay
py = df.loc['Paraguay']

'''
    ##### function ######
'''


def nlargesmall_est(largest, serie, n):
    # py value
    py = int(serie.loc['Paraguay'])
    x = serie.to_dict()
    # sort dict
    if largest is True:
        # descending order
        sorted_x = sorted(x.items(), key=operator.itemgetter(1), reverse=True)
    else:
        # ascending order
        sorted_x = sorted(x.items(), key=operator.itemgetter(1), reverse=False)
    # array for index
    index = np.empty(n, dtype=object)
    # array for the values
    values = np.empty(n, dtype=object)
    for i in range(n):
        index[i] = sorted_x[i][0]
        values[i] = sorted_x[i][1]
    # new ordered series
    s = pd.Series(data=values, index=index)
    tempy = pd.Series(data=[py], index=['Paraguay'])
    s = s.append(tempy)
    print(s)


'''
    ##### 4 smallest countries + py ######
'''
print('Los 4 países con menores valores de total de casos positivos. + Py')
nlargesmall_est(False, df, 4)


'''
    ##### 4 largest countries + py ######
'''
print('Los 4 países con mayores valores de total de casos positivos. + Py')
nlargesmall_est(True, df, 4)


'''
    ##### cases per million section ######
'''

# Read the file
df_locations = pd.read_csv("locations.csv", low_memory=False)
# get total cases for last row (last date) in file total_cases
df_cases = data.iloc[number_rows]
# make it a series
s_cases = pd.Series(data=df_cases, name='casos')
# series with same index but data is number of population
s_population = pd.Series(data=df_locations['population'].array, index=df_locations['location'].array, name='population')
# s_population = s_population.drop(columns='International')
# empty array to store result values
arr = []
for i in df_cases.index:
    arr.append(s_cases[i]*1000000/s_population[i])
# result series
s_cases_per_million = pd.Series(data=arr, index=df_locations['location'].array, name='casos por millon')
# round data
s_cases_per_million = s_cases_per_million.round()
# get number of total cases from paraguay
py = s_cases_per_million.loc['Paraguay']
tempy = pd.Series(data=[py], index=['Paraguay'])

'''
    ##### 4 largest cases per million + py ######
'''
s_population = s_population.drop(index='International')
s_cases_per_million = s_cases_per_million.drop(index='International')
print('los 4 países con mayores casos positivos por millón de habitantes. + Py')
nlargesmall_est(True, s_cases_per_million, 4)
'''
    ##### 4 smallest cases per million + py ######
'''
print('los 4 países con menores casos positivos por millón de habitantes. + Py')
nlargesmall_est(False, s_cases_per_million, 4)

'''
    ##### Ejercicio e , last 10 days average all countries ######
'''


def rounded_mean(series, n):
    x = series
    x = sum(x)
    mean = round(x/n)
    return mean


# get subset dataframe with last 11 rows to process
df_average = df_total_cases.tail(11)
# index numbers for getting dates
first_index = df_average.index[1]
last_index = df_average.index[-1]
# text getting dates
days_date = 'fecha inicio: ' + df_average['date'][first_index] + ' hasta: ' + df_average['date'][last_index]
# del date column
df_average = df_average.drop(columns='date')
print('\nPromedio diario de casos positivos de los últimos 10 días de todos los países.\n' + days_date + '\n')
# get index in form of list
index = df_average.index.tolist()
# pop first element, we need 11 records to process but result is 10 rows
index.pop(0)
# temporal series to save diary cases
s_diary_cases = pd.Series(index=index, dtype='float64')
# result series with mean for every country
s_mean = pd.Series(index=df_average.columns, dtype='float64', name='promedio')
for i in df_average.columns:
    for j in index:
        # calculate diary cases per country: date 2 diary cases = date 2 total cases - date 1 total cases
        s_diary_cases[j] = df_average[i][j] - df_average[i][j-1]
    # find de average per country
    s_mean[i] = rounded_mean(s_diary_cases.to_list(), len(s_diary_cases.index))
# print all rows
# pd.set_option('display.max_rows', None)
print(s_mean)

'''
    ##### Ejercicio f , correlation cases per day######
'''


# last 15 rows cases per day
df_corr = data.tail(16).diff()
# delete null row
df_corr = df_corr.drop(index=df_corr.index.start)
# pearson correlation formula, let be xi , yi individual sample point
# numerator is sum from n=1 to n ( ( xi - mean(x )*( yi - mean(y) ) )
# denominator is sqrt( std(x)*std(y) ) where st(x) d is sum from i = i to n ( xi - mean(x) )^2


def correlation(x, y):
    # this calculates correlation between to data arrays
    x = x
    n = len(x)
    # Finding the mean of the series x and y
    mean_x = sum(x)/float(len(x))
    mean_y = sum(y) / float(len(y))
    cov = 0
    std_x = 0
    std_y = 0
    for i in range(n):
        cov += (x[i]-mean_x)*(y[i]-mean_y)
        std_x += (x[i]-mean_x)**2
        std_y += (y[i] - mean_y) ** 2
    std_x = std_x**0.5
    std_y = std_y ** 0.5
    denominator = std_x*std_y

    def safe_div(numerator, denominator):
        if denominator == 0:
            return 0
        else:
            return numerator / denominator

    return safe_div(cov, denominator)


def corr(df, sample):
    # apply correlation to every column
    # columns name to iterate
    columns_name = df.columns.to_list()
    # series that store correlation
    s_result = pd.Series(index=columns_name, dtype='float64')
    for i in columns_name:
        # calculate correlation for every country
        s_result[i] = correlation(df[i].array, df[sample].array)
    return s_result


s_corr = corr(df_corr, 'Paraguay')
# hide paraguay result when doing nlargest
s_corr['Paraguay'] = -1
print('Los países que tienen la mayor correlación en los últimos 15 días, con respecto a Paraguay.')
print(s_corr.nlargest(10))

'''
    ##### Ejercicio g , grafico ######
'''
arr = ['Paraguay', 'Brazil', 'Argentina', 'Bolivia', 'Uruguay']
df_plot = pd.DataFrame(dtype='float64')
for i in arr:
    df_plot[i] = df_total_cases[i]
df_plot = df_plot.diff()
df_plot.insert(loc=0, column='date', value=df_total_cases['date'])
df_plot.plot(x='date', y=arr)
plt.show()

